### PR TITLE
API Authentication

### PR DESCRIPTION
**Pull request by htjahyon**
image

### CODE REVIEW

File: api-authentication/server/index.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    throw new ClientError(401, 'invalid login');
+  }
+
+  /* your code starts here */
+  let usernameData, hashedPasswordData, userId;
+  const sql = `
+        select "username",
+               "hashedPassword"
+        from "users"
+        where "hashedPassword" = $2
```
Comment (scbowler) [inline api-authentication/server/index.js]: You want to get the user by their `username`, should be `username = $1`

File: api-authentication/server/index.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        returning "userId", "username", "password"`;
```
Comment (scbowler) [inline api-authentication/server/index.js]: You don't need a `returning` on a `select` statement, that is only for `insert` and `update`

File: api-authentication/server/index.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  const params = [userId, usernameData, hashedPasswordData];
```
Comment (scbowler) [inline api-authentication/server/index.js]: The only param you need in your array is the user's `username`

File: api-authentication/server/index.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      }
+      db.query(sql, params)
+        .then(result => {
+          const payload = { userId, username };
+          const token = jwt.sign(payload, process.env.TOKEN_SECRET);
+          res.status(200).json(token);
+        })
+        .catch(err => next(err));
+    })
+    .catch(err => next(err));
```
Comment (scbowler) [inline api-authentication/server/index.js]: You need to query the database first to get the `hashedPassword` from the database. Then you use `.verify` to compare the `hashedPassword` to the `password` the user sent to the endpoint.

**Review by scbowler — CHANGES_REQUESTED** [review]
Some updates needed, see below comments for details. Let me know if you have any questions.

**Comment by htjahyon** [comment]
image

**Review by scbowler — APPROVED** [review]
Great job implementing the feedback! Looks great! 👍

**Pull request merged**